src/leo_nlm_proto4_server.erl

```diff
@@ -0,0 +1,314 @@
+-define(NLM4_ROFS,                  'NLM4_ROFS').
+-define(NLM4_STALE_FH,              'NLM4_STALE_FH').
+-define(NLM4_FBIG,                  'NLM4_FBIG').
+-define(NLM4_FAILED,                'NLM4_FAILED').
+
+%% ---------------------------------------------------------------------
+%% API
+%% ---------------------------------------------------------------------
+%% @doc Called only once from a parent rpc server process to initialize this module
+%%      during starting a leo_storage server.
+-spec(init(any()) -> {ok, any()}).
+init(_Args) ->
+    ets:new(?NLM_LOCK_ETS, [set, named_table, public]),
+    {ok, void}.
+
+handle_call(Req, _From, S) ->
+    ?debug("handle_call", "req:~p from:~p", [Req, _From]),
+    {reply, [], S}.
+
+handle_cast(Req, S) ->
+    ?debug("handle_cast", "req:~p", [Req]),
+    {reply, [], S}.
+
+handle_info(Req, S) ->
+    ?debug("handle_info", "req:~p", [Req]),
+    {noreply, S}.
+
+terminate(_Reason, _S) ->
+    ok.
+
+%% ---------------------------------------------------------------------
+%% API
+%% ---------------------------------------------------------------------
+%% @doc
+nlmproc4_null_4(_Clnt, State) ->
+    ?debug("NLM_NULL", "from:~p", [_Clnt]),
+    {reply, [], State}.
+
+%% @doc
+nlmproc4_test_4({Cookie, IsExclusive, Lock} = TestArgs, Clnt, State) ->
+    ?debug("NLM_TEST", "Args:~p, from:~p", [TestArgs, Clnt]),
+    {_CallerName, FileHandler, _Owner, _Svid, _Offset, _Length} = Lock,
+    NewLock = convert_lock(Lock, IsExclusive), 
+    LockRet = case test_lock(FileHandler, NewLock) of
+                  {ok, _} ->
+                      ?NLM4_GRANTED;
+                  _ ->
+                      ?NLM4_DENIED
+              end,
+    {reply, {Cookie, {LockRet}}, State}.
+
+nlmproc4_lock_4({Cookie, _IsBlock, IsExclusive, Lock, _IsReclaim, _NSMState} = LockArgs, Clnt, State) ->
+    ?debug("NLM_LOCK", "Args:~p, from:~p", [LockArgs, Clnt]),
+    {_CallerName, FileHandler, _Owner, _Svid, _Offset, _Length} = Lock,
+    NewLock = convert_lock(Lock, IsExclusive), 
+    LockRet = case test_lock(FileHandler, NewLock) of
+                  {ok, ExistingLocks} ->
+                      ets:insert(?NLM_LOCK_ETS, {FileHandler, [NewLock | ExistingLocks]}),
+                      ?NLM4_GRANTED;
+                  _ ->
+                      ?NLM4_DENIED
+              end,
+    {reply, {Cookie, {LockRet}}, State}.
+
+nlmproc4_cancel_4({Cookie, _IsBlock, _IsExclusive, _Lock} = CancArgs, Clnt, State) ->
+    ?debug("NLM_CANCEL", "Args:~p, from:~p", [CancArgs, Clnt]),
+    {reply, {Cookie, {?NLM4_GRANTED}}, State}.
+
+nlmproc4_unlock_4({Cookie, Lock} = UnlockArgs, Clnt, State) ->
+    ?debug("NLM_UNLOCK", "Args:~p, from:~p", [UnlockArgs, Clnt]),
+    {_CallerName, FileHandler, Owner, _Svid, Offset, Length} = Lock,
+    End = Offset + Length - 1,
+    case ets:lookup(?NLM_LOCK_ETS, FileHandler) of
+        [{FileHandler, ExistingLocks}] ->
+            NewLocks = lists:foldl(fun(CurLock, Acc) ->
```

Comment:
Since you can use pattern match at function arguments,
Being able to remove the outer case clause (`case CurLock#lock_record.owner of`)